src/ui/citation-popup/view.js

```diff
@@ -1,22 +1,107 @@
+    if (event.target.classList.contains("icon_remove")) {
+        removeReference(event.target.parentNode.id);
+    }
+});
 
-// When selecting a certain reference, set the "reference_id" field accordingly
-let selectedReferenceView = null;
-for (let i = 0; i < refs.length; i++) {
-    let ref = refs[i];
-    ref.addEventListener("click", event => {
-        referenceIdInput.value = event.target.id;
-        
-        // Attach some css style
-        if (selectedReferenceView !== null) {
-            selectedReferenceView.className = "";
+function configAutoComplete () {
+
+    const autoCompleteJS = new autoComplete({
+        placeHolder: "Search for references...",
+        data: {
+            src: refs,
+            keys: ["title"],
+            filter: list => {
+                return list.filter(item => !state.selectedRefs.has(item.value["id"]));
+            }
+        },
+        resultsList: {
+            noResults: true,
+            maxResults: 15,
+            tabSelect: true
+        },
+        resultItem: {
+            element: (item, data) => {
+                // Modify Results Item Style
+                item.style = "display: flex; justify-content: space-between;";
+                // Modify Results Item Content
+                item.innerHTML = `
+                    <span style="text-overflow: ellipsis; white-space: nowrap; overflow: hidden;">
+                        ${data.match}
+                    </span>
+                `;
+            },
+            highlight: true
+        },
+        events: {
+            input: {
+                focus: () => {
+                    if (autoCompleteJS.input.value.length) autoCompleteJS.start();
+                }
+            }
         }
+    });
+    autoCompleteJS.searchEngine = "loose";
 
-        selectedReferenceView = event.target;
-        selectedReferenceView.className = "selected";
+    autoCompleteJS.input.addEventListener("selection", event => {
+        const feedback = event.detail;
+        const selection = feedback.selection.value;
+        addReference(selection["id"]);
     });
+
+}
+
+function addReference (refId = "") {
+    state.selectedRefs.add(refId);
+    render();
+}
+
+function removeReference (refId = "") {
+    state.selectedRefs.delete(refId);
+    render();
+}
+
+/* Rendering state-based UI */
+function render () {
+    const selectedRefsArray = Array.from(state.selectedRefs);
+    selectedRefsView.innerHTML = template(selectedRefsArray);
+    output.value = JSON.stringify(selectedRefsArray);
+}
+
+/**
+ * Returns an HTML representation of an array of refs
+ * @param {Reference[]} refs
+ * @returns string
+ */
+function template (refs = []) {
+    if (refs.length === 0) {
+        return "Select some references to be added to the current note";
+    }
+    return (
+        refs
+            .map(refId => refsMap.get(refId))           // id => reference
+            .map(ref => (`
+                <li id="${ref["id"]}">
```

Comment:
ID needs to be escaped (even if now you know it's not needed, it's always safer to do so, for example if later on the ID format is changed).

Reply:
Is it necessary? I escape the JSON data before passing it to the view. Isn't that enough?

Reply:
Yes, escaping variables is not something optional to be added after later on. And I don't see what you mean by escaping the JSON data before passing it to the view

- First, if these properties are already escaped, then they are not named correctly. It's not just an "id" it's an "idHtml" - it's HTML code basically so it can't be used in an SQL query for example or other context that would need different escaping (see [Making Wrong Code Look Wrong](https://www.joelonsoftware.com/2005/05/11/making-wrong-code-look-wrong/) for good ideas on how to name variables)

- Secondly, you should only escape data at the point where it is needed. Pre-escaping data is a recipe for accidentally double-escaping, and often these issues are not obvious during development, only when people start importing all kind of unexpected data.

Somehow we keep going back to these escaping issues, so please try to remember to always escape data that you insert into strings. Perhaps do some reading about the topic to better understand how important it is. But to sum up:

- If you insert the data into HTML, use HTML entities encoding

- If you insert into a URL, use encodeURIComponent or encodeURI (you have the article where they explain the difference)